src/NServiceBus.RavenDB/Outbox/RavenDbOutboxStorage.cs

```diff
@@ -56,40 +56,42 @@ public OutboxCleaner(OutboxRecordsCleaner cleaner, TimeSpan frequencyToRunDedupl
                 this.timeToKeepDeduplicationData = timeToKeepDeduplicationData;
             }
 
-            protected override Task OnStart(IMessageSession session)
+            protected override Task OnStart(IMessageSession session, CancellationToken cancellationToken = default)
             {
                 if (frequencyToRunDeduplicationDataCleanup == Timeout.InfiniteTimeSpan)
                 {
                     return Task.CompletedTask;
                 }
 
-                cancellationTokenSource = new CancellationTokenSource();
-                cancellationToken = cancellationTokenSource.Token;
-
-                cleanupTask = Task.Run(() => PerformCleanup(), CancellationToken.None);
+                cleanupCancellationTokenSource = new CancellationTokenSource();
+                cleanupTask = Task.Run(() => PerformCleanup(cleanupCancellationTokenSource.Token), cancellationToken);
 
                 return Task.CompletedTask;
             }
 
-            protected override async Task OnStop(IMessageSession session)
+            protected override async Task OnStop(IMessageSession session, CancellationToken cancellationToken = default)
             {
-                cancellationTokenSource.Cancel();
+                cleanupCancellationTokenSource.Cancel();
 
                 if (cleanupTask == null)
                 {
                     return;
                 }
 
-                var timeoutTask = Task.Delay(TimeSpan.FromSeconds(30));
-                var finishedTask = await Task.WhenAny(cleanupTask, timeoutTask).ConfigureAwait(false);
+                var taskCompletionSource = new TaskCompletionSource<bool>();
 
-                if (finishedTask == timeoutTask)
+                using (cancellationToken.Register(() => taskCompletionSource.SetResult(true)))
                 {
-                    logger.Error("RavenOutboxCleaner failed to stop within the time allowed (30s).");
+                    var finishedTask = await Task.WhenAny(cleanupTask, taskCompletionSource.Task).ConfigureAwait(false);
+
+                    if (finishedTask == taskCompletionSource.Task)
+                    {
+                        logger.Error("RavenOutboxCleaner failed to stop within the time allowed (30s).");
```

Comment:
The 30 second task has gone, so this message seems wrong.

In fact, should the method throw an `OperationCancelledException` if the passed-in token is cancelled?